dd-sdk-android-gradle-plugin/src/main/kotlin/com/datadog/gradle/plugin/DdNdkSymbolFileUploadTask.kt

```diff
@@ -0,0 +1,191 @@
+                }
+                currentDir = currentDir.parentFile
+                levelsUp++
+            }
+            return null
+        }
+
+        @Suppress("MagicNumber", "ReturnCount")
+        private fun isAgp7OrAbove(): Boolean {
+            val version = Version.ANDROID_GRADLE_PLUGIN_VERSION
+            val groups = version.split(".")
+            if (groups.size < 3) return false
+            val major = groups[0].toIntOrNull()
+            val minor = groups[1].toIntOrNull()
+            val patch = groups[2].substringBefore("-").toIntOrNull()
+            if (major == null || minor == null || patch == null) return false
+            return major >= 7 && minor >= 0 && patch >= 0
+        }
+
+        private fun fixNativeOutputPath(taskFolder: File): File {
+            return taskFolder.parentFile.parentFile.takeIf { it.parentFile.name == "cxx" }
+                ?: taskFolder.parentFile.parentFile.parentFile.takeIf { it.parentFile.name == "cxx" }
+                ?: taskFolder
+        }
+
+        private fun getSearchDir(buildTask: TaskProvider<ExternalNativeBuildTask>, propertyName: String, providerFactory: ProviderFactory): Provider<File> {
+            return buildTask.flatMap { task ->
+                val soFolder = ExternalNativeBuildTask::class.memberProperties.find { it.name == propertyName }?.get(task)
+                when (soFolder) {
+                    is File -> providerFactory.provider { fixNativeOutputPath(soFolder) }
+                    is DirectoryProperty -> soFolder.map { fixNativeOutputPath(it.asFile) }
+                    else -> providerFactory.provider { File("") }
+                }
+            }
+        }
+
+        @Suppress("LongParameterList", "ReturnCount")
+        fun register(
+            project: Project,
+            variant: ApplicationVariant,
+            buildIdTask: TaskProvider<GenerateBuildIdTask>,
+            providerFactory: ProviderFactory,
+            apiKey: ApiKey,
+            extensionConfiguration: DdExtensionConfiguration,
+            repositoryDetector: RepositoryDetector
+        ): TaskProvider<DdNdkSymbolFileUploadTask>? {
+            val nativeBuildProviders = variant.externalNativeBuildProviders
+            if (nativeBuildProviders.isEmpty()) {
+                LOGGER.info("No native build tasks found for variant ${variant.name}, skipping symbol file upload.")
+                return null
+            }
+
+            return project.tasks.register(
+                TASK_NAME + variant.name.capitalize(),
+                DdNdkSymbolFileUploadTask::class.java
+            ) { task ->
+                val roots = mutableListOf<File>()
+                variant.sourceSets.forEach { it ->
+                    roots.addAll(it.javaDirectories)
+                    if (isAgp7OrAbove()) {
+                        roots.addAll(it.kotlinDirectories)
+                    }
+                }
+                task.sourceSetRoots = roots
```

Comment:
That is actually the thing we **need** to improve somehow (no idea how for now), because we will upload repository info several times (and this repo info may be big, big projects may have thousands of files). Ideally we need to make an upload only once, but with that approach we will do upload for JVM mapping file, plus for each `.so` file we found.

I would say we shouldn't upload repo info for NDK symbol files, because this info will be uploaded with JVM mapping file, but then we may have a situation when app has NDK symbols, but no obfuscation (hence no mapping file).

I guess it will be too complicated to address this in this PR, so better to create a task for the future (maybe even some backend changes will be needed).

Reply:
Ideally we need to scan Git repo only once as well, because scan is expensive, it involves `git` CLI call. But yeah, let's address it in the future tickets.

Reply:
Created as RUM-3171 and RUM-3172